### PR TITLE
[WFLY-13188] Defer vendor metric registration until completion of boot

### PR DESCRIPTION
**Pull request by bstansberry**
https://issues.redhat.com/browse/WFLY-13188

### CODE REVIEW

**Review by xstefank — APPROVED** [review]
I am no expert in this area. However, I don't see any issues with this PR.

File: microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricCollector.java [hunk 1]
```diff
@@ -88,6 +94,26 @@ public MetricRegistration collectResourceMetrics(final Resource resource,
...
+                    registration.unregister();
+                    processStateNotifier.removePropertyChangeListener(this);
+                }
+
+            }
+        };
+        this.processStateNotifier.addPropertyChangeListener(listener);
+        // If server is already running, we won't get a change event so register now
+        if (ControlledProcessState.State.RUNNING == this.processStateNotifier.getCurrentState()) {
+            registration.register();
```
Comment (spyrkob) [inline microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricCollector.java:115]: @bstansberry it's probably a stupid question, but is there a risk of registering the metric twice if the state flips to RUNNING after listener is added but before the condition?

Comment (bstansberry) [reply microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricCollector.java:115]: @spyrkob No, there isn't, but it's probably better to code defensively against it by making the method synchronized.

It's not a stupid question. :)

This method is called in the context of a management op that modifying the system. The switch to RUNNING also occurs in the context of such a management op (a big composite op that runs during boot) and the switch to RUNNING doesn't happen until the code that would call this method has completed. The system won't allow concurrent execution of 2 such management ops.

Comment (spyrkob) [reply microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricCollector.java:115]: Great, thanks for the explanation :)

**Review by spyrkob — APPROVED** [review]
As far as I can tell this looks good, but tbh I'm not very familiar with MP subsystem

File: microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricCollector.java (hunk 1 above)
Comment (bstansberry) [reply microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricCollector.java:115]: @spyrkob I added the synchronized

**Pull request merged**